docs/contribute/bmi.qmd

```diff
@@ -0,0 +1,46 @@
+`get_current_time()`      | Get the current time of the Ribasim simulation
+`get_end_time()`          | Get the final time of the Ribasim simulation in seconds
+`get_start_time()`        | Get the start time of the Ribasim simulation (`0.0`)
+`get_time_step()`         | Get the proposed next internal Ribasim timestep
+`get_time_units()`        | Get the time unit (`s`)
+`get_value_ptr(string)`   | Get the pointer to a Ribasim internal array (see below)
+`update()`                | Perform a Ribasim internal time step
+`update_until(time)`      | Set Ribasim internal timesteps until the specified time
+
+Depending on what is specified in the Ribasim TOML configuration file, Ribasim can internally have adaptive (non-constant) timesteps. `update_until` will always try to progress the Ribasim simulation to exactly the time specified. This however can fail for algorithms that only support a fixed timestep if that timestep does not fit into the interval until the specified time an integer amount of times.
+
+## Memory pointers
+
+The following pointers to memory containing Ribasim internal arrays are given via the BMI using `get_value_ptr(string)`:
+
+string                          | meaning                                | type    | unit         | temporal type         | writable  | sorted by
+------------------------------- | -------------------------------------- | ------- | ------------ | --------------------- | --------  |----------
+`basin.storage`                 | storage per basin                      | Float64 | $m^3$        | instantaneous         | no        | basin node ID
+`basin.level`                   | level per basin                        | Float64 | $m$          | instantaneous         | no        |  basin node ID
+`basin.infiltration`            | infiltration flux per basin            | Float64 | $m^3 s^{-1}$ | forward fill          | yes       |  basin node ID
+`basin.drainage`                | drainage flux per basin                | Float64 | $m^3 s^{-1}$ | forward fill          | yes       |  basin node ID
+`basin.infiltration_integrated` | cumulative infiltration per basin      | Float64 | $m^3$        | integrated from start | yes       |  basin node ID
+`basin.drainage_integrated`     | cumulative drainage per basin          | Float64 | $m^3$        | integrated from start | yes       |  basin node ID
+`basin.subgrid_level`           | subgrid level                          | Float64 | $m$          | instantaneous         | no        |  subgrid ID
+`user_demand.demand`            | demand per node ID per priority        | Float64 | $m^3 s^{-1}$ | forward fill          | yes       |  user_demand node ID, priority index
+`user_demand.realized`          | cumulative intake flow per user        | Float64 | $m^3$        | integrated from start | yes       |  user_demand node ID
+
+Additional notes:
+
+- `user_demand.demand` yields the only 2D array, the other arrays are 1D. This array is indexed as `(node_idx, priority_idx)`
```

Comment:
nb. if memory presented as a one-dimensional array, runs fastest over priorities, then over user nodes, (therefore in a column major view such as in Julia or Fortran presented as (node_idx, priority_idx), but in row major views such as in Python as (priority_idx, node_idx)